tf_encrypted/tensor/boolfactory.py

```diff
@@ -0,0 +1,330 @@
+
+    def _lift(x, y) -> Tuple['Tensor', 'Tensor']:
+
+        if isinstance(x, Tensor) and isinstance(y, Tensor):
+            return x, y
+
+        if isinstance(x, Tensor):
+            return x, x.factory.tensor(y)
+
+        if isinstance(y, Tensor):
+            return y.factory.tensor(x), y
+
+        raise TypeError("Don't know how to lift {} {}".format(type(x), type(y)))
+
+    class Tensor(AbstractTensor):
+        """Base class for other native tensor classes."""
+
+        @property
+        @abc.abstractproperty
+        def value(self):
+            pass
+
+        @property
+        @abc.abstractproperty
+        def shape(self):
+            pass
+
+        def identity(self):
+            value = tf.identity(self.value)
+            return DenseTensor(value)
+
+        def to_native(self) -> tf.Tensor:
+            return self.value
+
+        def __repr__(self) -> str:
+            return '{}(shape={})'.format(type(self), self.shape)
+
+        @property
+        def factory(self):
+            return FACTORY
+
+        @property
+        def dtype(self):
+            return self.factory.native_type
+
+        def __getitem__(self, slc):
+            return DenseTensor(self.value[slc])
+
+        def transpose(self, perm):
+            return DenseTensor(tf.transpose(self.value, perm))
+
+        def strided_slice(self, args, kwargs):
+            return DenseTensor(tf.strided_slice(self.value, *args, **kwargs))
+
+        def gather(self, indices: list, axis: int = 0):
+            return DenseTensor(tf.gather(self.value, indices, axis=axis))
+
+        def split(self, num_split: int, axis: int = 0):
+            values = tf.split(self.value, num_split, axis=axis)
+            return [DenseTensor(value) for value in values]
+
+        def reshape(self, axes: Union[tf.Tensor, List[int]]):
+            return DenseTensor(tf.reshape(self.value, axes))
+
+        def equal(self, other, factory=None):
+            x, y = _lift(self, other)
+            factory = factory or FACTORY
+            return factory.tensor(tf.cast(tf.equal(x.value, y.value),
+                                          dtype=factory.native_type))
+
+        def expand_dims(self, axis: Optional[int] = None):
+            return DenseTensor(tf.expand_dims(self.value, axis))
+
+        def squeeze(self, axis: Optional[List[int]] = None):
+            return DenseTensor(tf.squeeze(self.value, axis=axis))
+
+        def cast(self, factory):
+            return factory.tensor(self.value)
+
+        def __xor__(self, other):
+            return self.xor(other)
+
+        def xor(self, other):
+            x, y = _lift(self, other)
+            value = tf.math.logical_xor(x.value, y.value)
+            return DenseTensor(value)
+
+        def __and__(self, other):
+            return self.and_(other)
+
+        def and_(self, other):
+            # Because "and" is a keyword in Python, the naming "and_" follows the way how Python handles this:
```

Comment:
we should be sure to standardize this name (and other logical ops) with SecureNN, whether we follow this pointer or convert to existing SecureNN naming